src/docs/torque.md

```diff
@@ -207,112 +178,236 @@ type int32 generates 'TNode<Int32T>' constexpr 'int32_t';
+
+```torque
+extern class OrderedHashMap extends HashTable;
+```
+
+#### Shapes
+
+Defining a `shape` looks just like defining a `class` except that it uses the keyword `shape` instead of `class`. A `shape` is a subtype of JSObject representing a point-in-time arrangement of properties on that object (in spec-ese, these are "data properties" rather than "internal slots"). It does not have its own instance type. An object with a particular shape may change and lose that shape at any time because the object might go into dictionary mode and move all of its properties out to a separate backing store.
+
+#### Structs
+
+`struct`s are collections of data that can easily be passed around together. (Completely unrelated to the class named `Struct`.) Like classes, they can include macros that operate on the data. Unlike classes, they also support generics. The syntax looks similar to a class:
+
+```torque
+@export
+struct PromiseResolvingFunctions {
+  resolve: JSFunction;
+  reject: JSFunction;
+}
+
+struct ConstantIterator<T: type> {
+  macro Empty(): bool {
+    return false;
+  }
+  macro Next(): T labels _NoMore {
+    return this.value;
+  }
+
+  value: T;
+}
+```
+
+##### Struct annotations
+
+Any struct marked as `@export` will be included with a predictable name in the generated file `gen/torque-generated/csa-types-tq.h`. The name is prepended with `TorqueStruct`, so `PromiseResolvingFunctions` becomes `TorqueStructPromiseResolvingFunctions`.
+
+Struct fields can be marked as `const`, which means they shouldn't be written to. The entire struct can still be overwritten.
+
+##### Structs as class fields
+
+A struct may be used as the type of a class field. In that case, it represents packed, ordered data within the class (otherwise, structs have no alignment requirements). This is particularly useful for indexed fields in classes. As an example, `DescriptorArray` contains an array of three-value structs:
+
+```torque
+struct DescriptorEntry {
+  key: Name|Undefined;
+  details: Smi|Undefined;
+  value: JSAny|Weak<Map>|AccessorInfo|AccessorPair|ClassPositions;
+}
+
+extern class DescriptorArray extends HeapObject {
+  const number_of_all_descriptors: uint16;
+  number_of_descriptors: uint16;
+  raw_number_of_marked_descriptors: uint16;
+  filler16_bits: uint16;
+  enum_cache: EnumCache;
+  descriptors[number_of_all_descriptors]: DescriptorEntry;
+}
+```
+
+##### References and Slices
+
+`Reference<T>` and `Slice<T>` are special structs representing pointers to data held within heap objects. They both contain an object and an offset; `Slice<T>` also contains a length. Rather than constructing these structs directly, you can use special syntax: `&o.x` will create a `Reference` to the field `x` within the object `o`, or a `Slice` to the data if `x` is an indexed field. `Reference<T>` can be dereferenced with `*` or `->`, consistent with C++.
```

Comment:
This should probably also mention that Reference\<T\> is not used directly, instead it has two subtypes  MutableReference\<T\> and ConstReference\<T\>, which can be referred to using syntactic sugar: &T and const &T.